src/sentry/api/endpoints/group_hashes_split.py

```diff
@@ -211,123 +203,182 @@ def _unsplit_group(group: Group, hash: str, hierarchical_hashes: Optional[Sequen
-                    ),
-                ]
-            )
-            .set_where(
-                common_where
-                + [
-                    Condition(
-                        Function(
-                            "has",
-                            [
-                                Column("hierarchical_hashes"),
-                                materialized_hash,
-                            ],
-                        ),
-                        Op.EQ,
-                        1,
-                    ),
-                ]
-            )
-            .set_groupby([Column("hashes")])
-            .set_orderby([OrderBy(Column("latest_event_timestamp"), Direction.DESC)])
-        )
+    try:
+        for variant in event.get_grouping_variants().values():
+            if not isinstance(variant, ComponentVariant):
+                continue
 
-        for row in snuba.raw_snql_query(query)["data"]:
-            assert not row["hashes"] or row["hashes"][hash_pos] == materialized_hash
+            if variant.get_hash() == tree["parentId"]:
+                tree["parentLabel"] = variant.component.tree_label
 
-            event_id = row["event_id"]
-            event = eventstore.get_event_by_id(group.project_id, event_id)
+            if variant.get_hash() == tree["id"]:
+                tree["label"] = variant.component.tree_label
 
-            tree = {
-                "parentId": _get_checked(row["hashes"], parent_pos),
-                "id": materialized_hash,
-                "childId": _get_checked(row["hashes"], child_pos),
-                "eventCount": row["event_count"],
-                "latestEvent": serialize(event, user, EventSerializer()),
-            }
+            if variant.get_hash() == tree["childId"]:
+                tree["childLabel"] = variant.component.tree_label
 
-            rv.append(tree)
+    except Exception:
+        sentry_sdk.capture_exception()
 
-            if not row["hashes"]:
-                continue
 
-            try:
-                for variant in event.get_grouping_variants().values():
-                    if not isinstance(variant, ComponentVariant):
-                        continue
+def _render_trees(group: Group, user):
+    materialized_hashes = list(
+        {gh.hash for gh in GroupHash.objects.filter(project=group.project, group=group)}
+    )
 
-                    if variant.get_hash() == tree["parentId"]:
-                        tree["parentLabel"] = variant.component.tree_label
+    # Evaluates to the index of the last hash that is in materialized_hashes,
+    # or 1 otherwise.
+    #
+    # XXX(markus): This is quite horrible but Snuba SDK does not allow us to do
+    # arrayMax([<other function call>, ...]), i.e. it does not allow function
+    # calls in array literals. So instead of arrayMax([1, 2, 3]) we do
+    # greatest(1, greatest(2, 3)).
+    find_hash_expr = 1
+    for hash in materialized_hashes:
```

Comment:
This is going to be an n*m operation based on the number of hashes and the size of `hierarchical_hashes`. What kind of numbers are we talking about here?

Reply:
I'm trying to find out exact numbers but my estimate is that this value is below 10 99% of the time.

Reply:
In general we are already down the road of "materialized hashes have to fit into the query" so I think if we significantly exceed this value all the time we'll have to come up with something else yet again.

Reply:
I just had to rewrite this because a group in prod had 30+ hashes passed here, meaning we were hitting recursion limits. Now instead of `greatest(a, greatest(b, greatest(c, d)` we produce `greatest(greatest(a, b), greatest(c, d))` which is still odd but does reduce stack depth to log2 of hashes.